### PR TITLE
CBG-2023: Internal properties are now prevented from use

### PR DESCRIPTION
**Pull request by IsaacLambat**
CBG-2023

- Internal properties are prevented depending on if using blip, import, or the REST API.
- Intentionally left todo in code for future use

`_deleted` is not rejected from import.

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [x] `xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration/219/ `1 failed test now fixed`
- [x] `just run failed test that has been fixed` https://jenkins.sgwdev.com/job/SyncGateway-Integration/220/

### CODE REVIEW

File: db/blip_handler.go [hunk 1]
```diff
@@ -824,6 +835,10 @@ func (bh *blipHandler) handleRev(rq *blip.Message) (err error) {
 		RevID: revID,
 	}
 	newDoc.UpdateBodyBytes(bodyBytes)
+	err = validateBlipBody(newDoc.Body())
```
Comment (bbrks) [inline db/blip_handler.go]: This is going to be adding extra unmarshalling with `bodyBytes` via `newDoc.Body()` in all BLIP document writes, whether they contain an underscore property or not.

File: db/crud.go [hunk 2]
```diff
@@ -1050,6 +1060,13 @@ func (db *Database) PutExistingRevWithBody(docid string, body Body, docHistory [
 
 }
 
+func validateDocUpdate(body Body) error {
```
Comment (bbrks) [inline db/crud.go]: It's not clear from the function name why this is only chekcing `_sync` and not `_id`, `_rev`, etc.
Can you clarify by renaming the function to be specific to REST API body handling and state why those properties are allowed through in this case?

File: db/crud.go [hunk 3]
```diff
@@ -1621,19 +1645,18 @@ func (db *Database) documentUpdateFunc(docExists bool, doc *Document, allowImpor
 		return
 	}
 
-	// Marshal raw user xattrs for use in Sync Fn. If this fails we can bail out so we should do early as possible.
-	metaMap, err := doc.GetMetaMap(db.Options.UserXattrKey)
+	err = validateNewBody(newDoc.Body())
```
Comment (bbrks) [inline db/crud.go]: It's not obvious to me that newDoc.Body is already unmarshalled at this point.

It might still end up being more efficient to call GetDeepMutableBody and do the validation on that, since we know we have to do that work anyway for a valid doc. Doing it that way means it would only be 1 unmarshal instead of potentially 2.

File: db/import.go [hunk 4]
```diff
@@ -443,6 +443,27 @@ func (db *Database) backupPreImportRevision(docid, revid string) error {
...
+
+	if isPurged, ok := body[BodyPurged].(bool); ok && isPurged {
+		return base.ErrImportCancelledPurged
+	}
+
+	// TODO: Validate attachment data to ensure user is not setting invalid attachments
+
+	if body == nil {
+		return base.ErrEmptyDocument
+	}
```
Comment (bbrks) [inline db/import.go]: Can't this be done at the very top as an early return?

File: db/blip_handler.go [hunk 5]
```diff
@@ -768,6 +768,17 @@ type removalDocument struct {
...
+func validateBlipBody(body Body) error {
+	// Prevent disallowed internal properties from being used
+	disallowed := []string{base.SyncPropertyName, BodyId, BodyRev, BodyDeleted, BodyRevisions}
+	for _, prop := range disallowed {
+		if body[prop] != nil {
+			return base.HTTPErrorf(http.StatusBadRequest, "top-level property '"+prop+"' is a reserved internal property")
+		}
+	}
+	return nil
+}
```
Comment (bbrks) [inline db/blip_handler.go]: I'm not keen on having these validation functions spread throughout the codebase.
It makes it a little difficult to compare what properties are allowed in the 3 write code-paths. Could we move the functions to a common place?

File: db/crud.go [hunk 6]
```diff
@@ -1040,6 +1045,11 @@ func (db *Database) PutExistingRevWithBody(docid string, body Body, docHistory [
 	delete(body, BodyAttachments)
 	newDoc.UpdateBody(body)
 
+	err = validateDocUpdate(body)
+	if err != nil {
+		return nil, "", err
+	}
```
Comment (bbrks) [inline db/crud.go]: earlier exit

```suggestion
	err = validateDocUpdate(body)
	if err != nil {
		return nil, "", err
	}

	newDoc.UpdateBody(body)
```

File: rest/admin_api_test.go [hunk 7]
```diff
@@ -4570,3 +4570,150 @@ func TestReplicatorCheckpointOnStop(t *testing.T) {
...
+			docBody:     `{"_id": "id_mismatch"}`,
+			docID:       "completely_different_id",
+			expectError: true,
+		},
+		{
+			name:        "Rev match",
+			docBody:     `{"_rev": "1-ca9ad22802b66f662ff171f226211d5c", "rand": "439870"}`,
+			rev:         "1-ca9ad22802b66f662ff171f226211d5c",
+			expectError: false,
+		},
```
Comment (bbrks) [inline rest/admin_api_test.go]: Is it worth adding test cases for either/or body/parameter? Or is there is sufficient coverage elsewhere for those cases?

**Review by bbrks — CHANGES_REQUESTED** [review]
I have some concerns about unintentionally doing extra unmarshalling work with these changes. See inline comments

File: db/crud.go (hunk 3 above)
Comment (IsaacLambat) [reply db/crud.go]: Yep. It seems that it would have been unmarshalled at this point already in the REST API case, but not in the blip case (causing extra unmarshalling). Will change it back.

File: rest/admin_api_test.go (hunk 7 above)
Comment (IsaacLambat) [reply rest/admin_api_test.go]: Didn't add these test cases as they are tested throughout the code however these are not what other tests are intending to test, so will add these test cases just to keep it specific to this change.

File: db/validation.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+			return base.HTTPErrorf(http.StatusBadRequest, "top-level property '"+prop+"' is a reserved internal property therefore cannot be imported")
+		}
+	}
+	// TODO: Validate attachment data to ensure user is not setting invalid attachments
+
+	return nil
+}
+
+// validateBlipBody validates incoming blip rev bodies
+func validateBlipBody(rawBody []byte, doc *Document) error {
```
Comment (IsaacLambat) [inline db/validation.go:67]: Takes in `rawBody` to save a call to `doc.BodyBytes()`

File: db/validation.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	// Prevent disallowed internal properties from being used
+	disallowed := []string{base.SyncPropertyName, BodyId, BodyRev, BodyDeleted, BodyRevisions}
+	for _, prop := range disallowed {
+		// Only unmarshal if raw body contains the disallowed property
+		if bytes.Contains(rawBody, []byte(prop)) {
+			if doc.Body()[prop] != nil {
```
Comment (bbrks) [inline db/validation.go]: This would allow `"_rev": null`, which doesn't seem right. I'd suggest doing this instead so it doesn't matter what the value is.

```suggestion
			if _, ok := doc.Body()[prop]; ok {
```

File: db/validation.go (hunk 8 above)
Comment (bbrks) [reply db/validation.go:67]: Can you put this in the godoc comment so it's clear why the function takes both?

File: db/validation.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+	return nil
+}
+
+// validateBlipBody validates incoming blip rev bodies
+func validateBlipBody(rawBody []byte, doc *Document) error {
+	// Prevent disallowed internal properties from being used
+	disallowed := []string{base.SyncPropertyName, BodyId, BodyRev, BodyDeleted, BodyRevisions}
+	for _, prop := range disallowed {
+		// Only unmarshal if raw body contains the disallowed property
+		if bytes.Contains(rawBody, []byte(prop)) {
```
Comment (bbrks) [inline db/validation.go]: Would be nice to see if this can be a tiny bit more targeted to a JSON property to avoid false-positives. I think the best we can do here is wrapping the property in double quotes though, as the following colon could include whitespace.

For example, a string that happens to contain `_deleted_` (e.g. markdown) will match the Contains case but then fail the subsequent body check.

```suggestion
		if bytes.Contains(rawBody, []byte(`"`+prop+`"`)) {
```

File: db/validation.go [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+		return base.HTTPErrorf(http.StatusNotFound, "Document revision is not accessible")
+	}
+
+	// Reject bodies that contains the "_purged" property.
+	if body[BodyPurged] != nil {
+		return base.HTTPErrorf(http.StatusBadRequest, "user defined top-level property '_purged' is not allowed in document body")
+	}
+
+	for key := range body {
+		if strings.HasPrefix(key, BodyInternalPrefix) {
```
Comment (bbrks) [inline db/validation.go:25]: I think this is relevant even for existing code doing this byte/string checking upfront - but these checks are case-sensitive, and the Go JSON implementation is case _insensitive._

That difference might be the cause of subtle bugs, if a user is able to bypass the validation by writing properties in uppercase.

File: rest/blip_api_test.go [hunk 12]
```diff
@@ -3975,3 +3975,131 @@ func TestAttachmentWithErroneousRevPos(t *testing.T) {
...
+			expectReject: false,
+		},
+		{
+			name:         "Invalid _sync",
+			inputBody:    map[string]interface{}{"_sync": true},
+			expectReject: true,
+		},
+		{
+			name:         "Valid _id",
+			inputBody:    map[string]interface{}{"_id": "documentid"},
```
Comment (bbrks) [inline rest/blip_api_test.go:4004]: As mentioned previously, our validation checks are mostly case sensitive. Would be interesting to include some uppercase tests for coverage.

**Review by bbrks — CHANGES_REQUESTED** [review]
I've thought of a potential edge case in the existing and new code that is doing byte/string Contains checks prior to unmarshalling. See inline comments. I think we'll need to figure out a way of addressing that.

File: db/validation.go (hunk 11 above)
Comment (IsaacLambat) [reply db/validation.go:25]: Will do as part of [CBG-2051](https://issues.couchbase.com/browse/CBG-2051)

File: rest/blip_api_test.go (hunk 12 above)
Comment (IsaacLambat) [reply rest/blip_api_test.go:4004]: Will add as part of [CBG-2051](https://issues.couchbase.com/browse/CBG-2051)

**Review by bbrks — APPROVED** [review]
